File: hello-grpc-python/conn/error_mapper.py
"""
Error mapper for translating gRPC status codes to human-readable messages
and implementing retry logic with exponential backoff.
"""

import asyncio
import logging
import time
from typing import Callable, Dict, Optional, TypeVar, Any
from dataclasses import dataclass

import grpc

logger = logging.getLogger(__name__)

T = TypeVar('T')


@dataclass
class RetryConfig:
    """Configuration for retry logic"""
    max_retries: int = 3
    initial_delay: float = 2.0  # seconds
    max_delay: float = 30.0  # seconds
    multiplier: float = 2.0

    @staticmethod
    def default() -> 'RetryConfig':
        """Creates default retry configuration"""
        return RetryConfig()


def map_grpc_error(error: Exception) -> str:
    """
    Maps gRPC status codes to human-readable error messages
    
    Args:
        error: The exception to map
        
    Returns:
        Human-readable error message
    """
    if error is None:
        return "Success"
    
    if isinstance(error, grpc.RpcError):
        code = error.code()
        message = error.details() if hasattr(error, 'details') else str(error)
        description = _get_status_description(code)
        
        if message:
            return f"{description}: {message}"
        return description
    
    return f"Unknown error: {str(error)}"


def _get_status_description(code: grpc.StatusCode) -> str:
    """
    Gets human-readable description for a gRPC status code
    
    Args:
        code: The status code
        
    Returns:
        Human-readable description
    """
    descriptions = {
        grpc.StatusCode.OK: "Success",
        grpc.StatusCode.CANCELLED: "Operation cancelled",
        grpc.StatusCode.UNKNOWN: "Unknown error",
        grpc.StatusCode.INVALID_ARGUMENT: "Invalid request parameters",
        grpc.StatusCode.DEADLINE_EXCEEDED: "Request timeout",
        grpc.StatusCode.NOT_FOUND: "Resource not found",
        grpc.StatusCode.ALREADY_EXISTS: "Resource already exists",
        grpc.StatusCode.PERMISSION_DENIED: "Permission denied",
        grpc.StatusCode.RESOURCE_EXHAUSTED: "Resource exhausted",
        grpc.StatusCode.FAILED_PRECONDITION: "Precondition failed",
        grpc.StatusCode.ABORTED: "Operation aborted",
        grpc.StatusCode.OUT_OF_RANGE: "Out of range",
        grpc.StatusCode.UNIMPLEMENTED: "Not implemented",
        grpc.StatusCode.INTERNAL: "Internal server error",
        grpc.StatusCode.UNAVAILABLE: "Service unavailable",
        grpc.StatusCode.DATA_LOSS: "Data loss",
        grpc.StatusCode.UNAUTHENTICATED: "Authentication required",
    }
    return descriptions.get(code, "Unknown error code")


def is_retryable_error(error: Exception) -> bool:
    """
    Determines if an error should be retried
    
    Args:
        error: The exception to check
        
    Returns:
        True if the error is retryable, False otherwise
    """
    if error is None:
        return False
    
    if not isinstance(error, grpc.RpcError):
        return False
    
    code = error.code()
    return code in (
        grpc.StatusCode.UNAVAILABLE,
        grpc.StatusCode.DEADLINE_EXCEEDED,
        grpc.StatusCode.RESOURCE_EXHAUSTED,
        grpc.StatusCode.INTERNAL,
    )


def handle_rpc_error(error: Exception, operation: str, context: Optional[Dict[str, Any]] = None) -> None:
    """
    Handles RPC errors with logging and context
    
    Args:
        error: The exception that occurred
        operation: The operation name
        context: Additional context information
    """
    if error is None:
        return
    
    error_msg = map_grpc_error(error)
    log_context = context.copy() if context else {}
    log_context['operation'] = operation
    log_context['error'] = error_msg
    
    if is_retryable_error(error):
        logger.warning(f"Retryable error occurred: {log_context}")
    else:
        logger.error(f"Non-retryable error occurred: {log_context}")


def retry_with_backoff(
    operation: str,
    func: Callable[[], T],
    config: Optional[RetryConfig] = None
) -> T:
    """
    Executes a function with exponential backoff retry logic (synchronous)
    
    Args:
        operation: The operation name for logging
        func: The function to execute
        config: The retry configuration
        
    Returns:
        The result of the function
        
    Raises:
        Exception: If all retries are exhausted
    """
    if config is None:
        config = RetryConfig.default()
    
    last_error = None
    delay = config.initial_delay
    
    for attempt in range(config.max_retries + 1):
        if attempt > 0:
            logger.info(
                f"Retry attempt {attempt}/{config.max_retries} for {operation} "
                f"after {delay}s"
            )
            time.sleep(delay)
        
        try:
            result = func()
            if attempt > 0:
                logger.info(f"Operation {operation} succeeded after {attempt + 1} attempts")
            return result
        except Exception as e:
            last_error = e
            
            if not is_retryable_error(e):
                logger.warning(f"Non-retryable error for {operation}: {map_grpc_error(e)}")
                raise
            
            if attempt < config.max_retries:
                # Calculate next delay with exponential backoff
                delay = min(delay * config.multiplier, config.max_delay)
    
    logger.error(
        f"Operation {operation} failed after {config.max_retries + 1} attempts: "
        f"{map_grpc_error(last_error)}"
    )
    raise Exception(f"Max retries exceeded for {operation}") from last_error


async def retry_with_backoff_async(
    operation: str,
    func: Callable[[], T],
    config: Optional[RetryConfig] = None
) -> T:
    """
    Executes a function with exponential backoff retry logic (asynchronous)
    
    Args:
        operation: The operation name for logging
        func: The async function to execute
        config: The retry configuration
        
    Returns:
        The result of the function
        
    Raises:
        Exception: If all retries are exhausted
    """
    if config is None:
        config = RetryConfig.default()
    
    last_error = None
    delay = config.initial_delay
    
    for attempt in range(config.max_retries + 1):
        if attempt > 0:
            logger.info(
                f"Retry attempt {attempt}/{config.max_retries} for {operation} "
                f"after {delay}s"
            )
            await asyncio.sleep(delay)
        
        try:
            result = await func()
            if attempt > 0:
                logger.info(f"Operation {operation} succeeded after {attempt + 1} attempts")
            return result
        except Exception as e:
            last_error = e
            
            if not is_retryable_error(e):
                logger.warning(f"Non-retryable error for {operation}: {map_grpc_error(e)}")
                raise
            
            if attempt < config.max_retries:
                # Calculate next delay with exponential backoff
                delay = min(delay * config.multiplier, config.max_delay)
    
    logger.error(
        f"Operation {operation} failed after {config.max_retries + 1} attempts: "
        f"{map_grpc_error(last_error)}"
    )
    raise Exception(f"Max retries exceeded for {operation}") from last_error

File: hello-grpc-php/common/utils/ErrorMapper.php
<?php

namespace Common\Utils;

use Grpc\Status;
use Exception;

/**
 * Configuration for retry logic
 */
class RetryConfig
{
    public int $maxRetries;
    public float $initialDelay; // seconds
    public float $maxDelay; // seconds
    public float $multiplier;

    public function __construct(
        int $maxRetries = 3,
        float $initialDelay = 2.0,
        float $maxDelay = 30.0,
        float $multiplier = 2.0
    ) {
        $this->maxRetries = $maxRetries;
        $this->initialDelay = $initialDelay;
        $this->maxDelay = $maxDelay;
        $this->multiplier = $multiplier;
    }

    public static function default(): self
    {
        return new self();
    }
}

/**
 * Error mapper for translating gRPC status codes to human-readable messages
 * and implementing retry logic with exponential backoff.
 */
class ErrorMapper
{
    /**
     * Maps gRPC status codes to human-readable error messages
     *
     * @param Exception|null $error The error to map
     * @return string Human-readable error message
     */
    public static function mapGrpcError(?Exception $error): string
    {
        if ($error === null) {
            return 'Success';
        }

        $code = $error->getCode();
        $message = $error->getMessage();
        $description = self::getStatusDescription($code);

        if (!empty($message)) {
            return "$description: $message";
        }
        return $description;
    }

    /**
     * Gets human-readable description for a gRPC status code
     *
     * @param int $code The status code
     * @return string Human-readable description
     */
    private static function getStatusDescription(int $code): string
    {
        $descriptions = [
            Status::OK => 'Success',
            Status::CANCELLED => 'Operation cancelled',
            Status::UNKNOWN => 'Unknown error',
            Status::INVALID_ARGUMENT => 'Invalid request parameters',
            Status::DEADLINE_EXCEEDED => 'Request timeout',
            Status::NOT_FOUND => 'Resource not found',
            Status::ALREADY_EXISTS => 'Resource already exists',
            Status::PERMISSION_DENIED => 'Permission denied',
            Status::RESOURCE_EXHAUSTED => 'Resource exhausted',
            Status::FAILED_PRECONDITION => 'Precondition failed',
            Status::ABORTED => 'Operation aborted',
            Status::OUT_OF_RANGE => 'Out of range',
            Status::UNIMPLEMENTED => 'Not implemented',
            Status::INTERNAL => 'Internal server error',
            Status::UNAVAILABLE => 'Service unavailable',
            Status::DATA_LOSS => 'Data loss',
            Status::UNAUTHENTICATED => 'Authentication required',
        ];

        return $descriptions[$code] ?? 'Unknown error code';
    }

    /**
     * Determines if an error should be retried
     *
     * @param Exception|null $error The error to check
     * @return bool True if the error is retryable, false otherwise
     */
    public static function isRetryableError(?Exception $error): bool
    {
        if ($error === null) {
            return false;
        }

        $code = $error->getCode();
        $retryableCodes = [
            Status::UNAVAILABLE,
            Status::DEADLINE_EXCEEDED,
            Status::RESOURCE_EXHAUSTED,
            Status::INTERNAL,
        ];

        return in_array($code, $retryableCodes, true);
    }

    /**
     * Handles RPC errors with logging and context
     *
     * @param Exception|null $error The error that occurred
     * @param string $operation The operation name
     * @param array $context Additional context information
     */
    public static function handleRpcError(?Exception $error, string $operation, array $context = []): void
    {
        if ($error === null) {
            return;
        }

        $errorMsg = self::mapGrpcError($error);
        $logContext = array_merge($context, [
            'operation' => $operation,
            'error' => $errorMsg,
        ]);

        $logger = LoggingConfig::getLogger();
        $contextStr = json_encode($logContext);

        if (self::isRetryableError($error)) {
            $logger->warning("Retryable error occurred: $contextStr");
        } else {
            $logger->error("Non-retryable error occurred: $contextStr");
        }
    }

    /**
     * Executes a function with exponential backoff retry logic
     *
     * @param string $operation The operation name for logging
     * @param callable $func The function to execute
     * @param RetryConfig|null $config The retry configuration
     * @return mixed The result of the function
     * @throws Exception If all retries are exhausted
     */
    public static function retryWithBackoff(string $operation, callable $func, ?RetryConfig $config = null)
    {
        if ($config === null) {
            $config = RetryConfig::default();
        }

        $lastError = null;
        $delay = $config->initialDelay;
        $logger = LoggingConfig::getLogger();

        for ($attempt = 0; $attempt <= $config->maxRetries; $attempt++) {
            if ($attempt > 0) {
                $delayMs = (int)($delay * 1000);
                $logger->info(
                    "Retry attempt $attempt/{$config->maxRetries} for $operation after {$delayMs}ms"
                );
                usleep($delayMs * 1000); // usleep takes microseconds
            }

            try {
                $result = $func();
                if ($attempt > 0) {
                    $attempts = $attempt + 1;
                    $logger->info("Operation $operation succeeded after $attempts attempts");
                }
                return $result;
            } catch (Exception $e) {
                $lastError = $e;

                if (!self::isRetryableError($e)) {
                    $errorMsg = self::mapGrpcError($e);
                    $logger->warning("Non-retryable error for $operation: $errorMsg");
                    throw $e;
                }

                if ($attempt < $config->maxRetries) {
                    // Calculate next delay with exponential backoff
                    $delay = min($delay * $config->multiplier, $config->maxDelay);
                }
            }
        }

        $attempts = $config->maxRetries + 1;
        $errorMsg = self::mapGrpcError($lastError);
        $logger->error("Operation $operation failed after $attempts attempts: $errorMsg");
        
        throw new Exception("Max retries exceeded for $operation", 0, $lastError);
    }
}
